### ISSUE TITLE
[Keras] Map and reduce functions.

### ISSUE THREAD

**Issue by shivarajugowda** [apache/incubator-mxnet#4162]
Does MXNET support higher order map-reduce kind of functions(map, foldl, foldr), like described below in TensorFlow
https://www.tensorflow.org/versions/r0.11/api_docs/python/functional_ops.html
Theno has similar functionality described here
http://deeplearning.net/software/theano/library/scan.html

If yes, any pointers on where to look is appreciated,
if not, any pointers on how to get started implementing them would be helpful. -Thanks.

**Comment by piiswrong** [comment]
what function are you trying to implement?
MXNet has a different programming model. Graphs are static and loops are done imperatively by calling executor.forward()/backward() and call imperative apis on the result

**Comment by shivarajugowda** [comment]
@piiswrong let me come up with the exact code and functions I need. I will get back to you.

**Comment by shivarajugowda** [comment]
@piiswrong is there a way in MXNET to implement element-wise version of java/C++ ternary operator ?
`x = a <= p ? 1 : 0;`
To give you a specific example I am looking to implement random_binomial function using the available random_uniform function on the same lines as [tensor_flow implementation](https://github.com/fchollet/keras/blob/master/keras/backend/tensorflow_backend.py) (search for random_binomial). In TensorFlow we can achieve such an operation using [tf.select operator](https://www.tensorflow.org/versions/r0.12/api_docs/python/control_flow_ops.html#select)

**Comment by shivarajugowda** [comment]
In some sense I am in a similar boat as @jsusskin in #1394. I am unable to find some of the aggregation operations(mean, std, prod, any, argmax, argmin) expecially per axis based ones. I understand there might be some indirect way of achieving them. But I think it would be clean to provide them in the API so that end-users do not have to reinvent them. I can help out with implementing them but just want to make sure I didn't miss something or if there is already an ongoing effort for it.

**Comment by piiswrong** [comment]
We have sum and max argmax in nnvm branch
We don't have binary compare ops but it's very easy to add those in src
operator tensor binary broadcast op

Shiv Gowda <notifications@github.com>于2016年12月10日 周六下午5:06写道：

**Comment by shivarajugowda** [comment]
OK, Thanks for the confirmation. I think implementing one of the compare ops might be a good starting point for me, I will work on it. Any idea on the timeline of the merge of NNVM branch to the master?

**Comment by piiswrong** [comment]
We just need more testing hopefully some time this month

Shiv Gowda <notifications@github.com>于2016年12月10日 周六下午7:28写道：

**Comment by shivarajugowda** [comment]
@piiswrong I have a pull request for the comparison operators #4182, please review when you get a chance.

**Comment by shivarajugowda** [comment]
@piiswrong is anybody pursuing Theano's Scan operator equivalent, is it in the pipeline or is there an equivalent existing mechanism?

**Comment by piiswrong** [comment]
We talked about it. It would be equivalent to a subgraph loop in mxnet. I don't think anyone is working on it now since you can do the same thing with imperative api.

We can have it just to make conversion from theano easier but I think scan was designed to workaround theano's limitations (lack of imerative api) and in the long term people should change the way they think and use imperative api instead

**Comment by piiswrong** [comment]
Actually I think you can easily simulate it with custom op

**Comment by shivarajugowda** [comment]
Thanks for the quickly reply. Let me chew on that.

**Comment by piiswrong** [comment]
for example, for rnn we use bucketing instead of scan. basically you create a graph for each k (length of sequence) and they share memory with each other. Then you just pick the one with the right length instead of looping inside a graph

**Comment by shivarajugowda** [comment]
@piiswrong I see that MXSymbolGrad is not yet implemented in NNVM branch. Is it in the works?

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
